### PR TITLE
tests: print all the serial logs for the nested test

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This PR includes:
 . Save logs when the nested service is restarted
 . Print all the logs created for debug for the current test
 . Move to the project env the nested env vars which don't change for the different nested projects  

The logs are deleted during the restore of each test, the idea of this is because during a test sometimes we start and stop the nested-vm service and when we start it the log from the previous start is overwritten. 
Currently we stop the vm to save image and we restart it, so in that case we are lossing the serial log for the first boot.

### CODE REVIEW

File: tests/lib/nested.sh [hunk 1]
```diff
@@ -679,7 +693,7 @@ nested_start_core_vm_unit() {
         if [ "$NESTED_ENABLE_SECURE_BOOT" = "true" ]; then
             cp -f "/usr/share/OVMF/OVMF_VARS.$OVMF_VARS.fd" "$NESTED_ASSETS_DIR/OVMF_VARS.$OVMF_VARS.fd"
             PARAM_BIOS="-drive file=/usr/share/OVMF/OVMF_CODE.$OVMF_CODE.fd,if=pflash,format=raw,unit=0,readonly -drive file=$NESTED_ASSETS_DIR/OVMF_VARS.$OVMF_VARS.fd,if=pflash,format=raw"
-            PARAM_MACHINE="-machine q35${ATTR_KVM} -global ICH9-LPC.disable_s3=1"
+            PARAM_MACHINE="-machine q35${ATTR_KVM},smm=on,mem-merge=off -global ICH9-LPC.disable_s3=1"
```
Comment (mvo5) [inline tests/lib/nested.sh:714]: This change seems unrelated to the commit message. Is this on purpose? If so I think putting it into it's own commit (with an explaination) would be better.

File: tests/lib/nested.sh [hunk 2]
```diff
@@ -588,6 +588,25 @@ nested_configure_cloud_init_on_core20_vm() {
...
+                cp "${NESTED_LOGS_DIR}/serial.log" "${NESTED_LOGS_DIR}/serial.log.${i}"
+                break
+            fi
+        done
+    fi
+}
+
+nested_print_serial_log() {
+    if [ -f "${NESTED_LOGS_DIR}/serial.log" ]; then
+        for logfile in $(ls -r "${NESTED_LOGS_DIR}"/serial.log*); do
```
Comment (bboozzoo) [inline tests/lib/nested.sh:604]: The logs are quite long, given that there can be many, it will be very hard to follow. Given that we're stating a VM for each test, I'd rather have just the serial output of a test that run last (and failed if used in debug section).

Comment (bboozzoo) [reply tests/lib/nested.sh:604]: There's a fix proposed in https://github.com/snapcore/snapd/pull/9352/commits/fbb76cb4c81617449a64a0acaf26d3b41e968837

Comment (sergiocazzolato) [reply tests/lib/nested.sh:604]: The logs are deleted during the restore of each test, the idea of this is because during a test sometimes we start and stop the nested-vm service and when we start it the log from the previous start is overwritten. 
Currently we stop the vm to save image and we restart it, so in that case we are lossing the serial log for the first boot.

File: tests/lib/nested.sh (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/nested.sh:714]: sure, thanks for the review

File: tests/lib/nested.sh [hunk 3]
```diff
@@ -613,6 +613,31 @@ nested_configure_cloud_init_on_core20_vm() {
...
+    fi
+}
+
+nested_print_serial_log() {
+    if [ -f "${NESTED_LOGS_DIR}/serial.log" ]; then
+        # here we disable SC2045 because previously it is checked there is at least
+        # 1 file which matches. In this case ls command is needed because it is important
+        # to get the list in reverse order.
+        # shellcheck disable=SC2045
+        for logfile in $(ls -r "${NESTED_LOGS_DIR}"/serial.log*); do
```
Comment (bboozzoo) [inline tests/lib/nested.sh:635]: If we have more than 10 files, the order will be somewhat confusing. From little testing this is the order we get with `ls -r`:

```
serial.log.9
serial.log.8
serial.log.7
serial.log.6
serial.log.5
serial.log.4
serial.log.3
serial.log.2
serial.log.10
serial.log.1
serial.log
```

I think we should use `seq 9` in `nested_save_serial_log`, so that we don't run into this problem. Also note, that the most recent log will have the highest number, so the order is still confusing. Maybe we should do something like this to ensure that the logs are printed in the order from the oldest to most recent?

```
for logfile in "$(ls "$NESTED_LOGS_DIR"/serial.log.*)"; do
    cat "$logfile"
done
cat "$NESTED_LOGS_DIR"/serial.log || true
```

Comment (sergiocazzolato) [reply tests/lib/nested.sh:635]: Sure, thanks for the suggestion

File: tests/lib/nested.sh [hunk 4]
```diff
@@ -682,6 +682,34 @@ nested_configure_cloud_init_on_core20_vm() {
     kpartx -d "$IMAGE"
 }
 
+nested_save_serial_log() {
+    if [ -f "${NESTED_LOGS_DIR}/serial.log" ]; then
+        for i in $(seq 50); do
```
Comment (bboozzoo) [inline tests/lib/nested.sh:687]: This should be `seq 1 9`, otherwise sorting done by `ls` is not so obvious anymore, eg, `1, 10, 2, 3, ..`.

Comment (sergiocazzolato) [reply tests/lib/nested.sh:687]: sorry, did that change but I lost that.

**Review by bboozzoo — APPROVED** [review]
LGTM, thanks.

File: spread.yaml [hunk 5]
```diff
@@ -54,6 +54,19 @@ environment:
     # Use the installed snapd and reset the systems without removing snapd
     REUSE_SNAPD: '$(HOST: echo "${SPREAD_REUSE_SNAPD:-0}")'
     PROFILE_SNAPS: '$(HOST: echo "${SPREAD_PROFILE_SNAPS:-0}")'
+
+    # Directory where the nested images and test assets are stored
+    NESTED_WORK_DIR: '$(HOST: echo "${NESTED_WORK_DIR:-/tmp/work-dir}")'
```
Comment (mvo5) [inline spread.yaml:59]: Thanks for cleaning this duplication :)
